modules/run-consul/run-consul

```diff
@@ -656,11 +318,59 @@ function run {
       "$redundancy_zone_tag" \
       "$disable_upgrade_migration" \
       "$upgrade_version_tag" \
+      "$enable_acl"
       "${recursors[@]}"
   fi
 
+  log_info "Creating Consul systemd file"
   generate_systemd_config "$SYSTEMD_CONFIG_PATH" "$config_dir" "$data_dir" "$systemd_stdout" "$systemd_stderr" "$bin_dir" "$user" "${environment[@]}"
+  log_info "Starting Consul"
   start_consul
+
+  if [[ "$enable_acl" == "true" ]]; then
+    local -r asg_name=$(aws_wrapper_get_asg_name $MAX_RETRIES $SLEEP_BETWEEN_RETRIES_SEC)
+    local -r aws_region=$(aws_get_instance_region)
+    local -r instance_id=$(aws_get_instance_id)
+
+    local bootstrap_token
+
+    # Calculate the rally point server for the ASG
+    local -r rally_point_hostname=$(aws_wrapper_get_asg_rally_point $asg_name $aws_region "false")
+    log_info "Calculated rally point instance is $rally_point_hostname."
+    local -r local_hostname=$(aws_wrapper_get_hostname)
+    log_info "Local hostname is $local_hostname"
+
+    if [[ "$rally_point_hostname" == "$local_hostname" ]]; then
+      log_info "Checking if bootstrap token already exists"
+      local -r existing_token=$(read_acl_token $cluster_tag_value "bootstrap" $aws_region "ssm" 1 0 "true")
+
+      if [[ "$existing_token" == "" ]] && [ "${server}" == "true" ]; then
+        log_info "Generating bootstrap ACL token"
+        bootstrap_token=$(generate_bootstrap_acl_token $MAX_RETRIES $SLEEP_BETWEEN_RETRIES_SEC)
+        log_info "Persisting bootstrap token to SSM parameter"
+        write_acl_token $bootstrap_token $cluster_tag_value "bootstrap" $aws_region "ssm"
+      else
+        log_info "Bootstrap token already exists, skipping"
+      fi      
+    fi
+
+    # If the bootstrap token isn't already read (i.e. if this is running on the non rally point node)
+    # then we need to read it.
+    if [[ -z "$bootstrap_token" ]]; then
+      log_info "Acquiring bootstrap token"
+      bootstrap_token=$(read_acl_token $cluster_tag_value "bootstrap" $aws_region "ssm")
+    fi
+
+    # Generate agent policy dynamically and write it
+    log_info "Creating agent policy and token"
+    local -r agent_policy=$(generate_node_acl_policy $local_hostname)
+    write_acl_policy "$instance_id" "$agent_policy" "$bootstrap_token"
+
+    # Generate agent token and persist it
+    local -r agent_token=$(generate_token "$instance_id" "$instance_id agent policy" $bootstrap_token)
```

Comment:
Did you know you can generate a token with -node-identity "$instance_id" and it will get the policy you want without creating it explicitly? (was looking because i just did something similar to this and yours is way better!)

Reply:
I did not know about that! It looks like it was added in Consul 1.8.1, so I wonder if it would be better to manually generate it since some folks might still be on an older version?

Reply:
hmm good point, hadnt realised it was as new as that.

any chance of switching on version/option-flag?  I found the node identity option useful to avoid having tons of mechanical acl policies to wade through: when every agent has its own explicit policy there ends up being tons

Reply:
We could make it so that if the Consul version is higher than 1.8.1 it will use that option, though I'd have to rethink the token generation function since that's specific to generating agent tokens (as opposed to the function now which is generic for any token given a policy). The bloat of agent policies is definitely a thing to consider, especially since ASG instances are not permanent things. Let me see how hard this would be, but I think we can include it here pretty easily.

Reply:
I've made this update though I wasn't able to test in a lower version because of some configuration challenges. Now it will use the new method when the version is 1.8.1 or higher. Thanks for the suggestion!